adam-core/src/main/scala/org/bdgenomics/adam/rdd/read/recalibration/DinucCovariate.scala

```diff
@@ -17,60 +17,98 @@
+import scala.annotation.tailrec
+
+/**
+ * An error covariate that tracks quality score estimation errors that are
+ * correlated with two nucleotides appearing in sequence in a read.
+ */
+private[adam] class DinucCovariate extends Covariate[(Char, Char)] {
+
+  /**
+   * @param read The read to compute the covariate for.
+   * @return Returns an array of dinucleotides.
+   */
+  def compute(read: AlignmentRecord): Array[(Char, Char)] = ComputingDinucCovariate.time {
+    val sequence = read.getSequence
+    if (read.getReadNegativeStrand) {
 
-// TODO: should inherit from something like AbstractCovariate[(DNABase, DNABase)]
-private[adam] class DinucCovariate extends AbstractCovariate[(Char, Char)] {
-  def compute(read: DecadentRead): Seq[Option[(Char, Char)]] = {
-    val sequence = read.residues.map(_.base)
-    if (read.isNegativeRead) {
       /* Use the reverse-complement of the sequence to get back the original
        * sequence as it was read by the sequencing machine. The sequencer
        * always reads from the 5' to the 3' end of each strand, but the output
        * from the aligner is always in the same sense as the reference, so we
        * use the reverse-complement if this read was originally from the
        * complementary strand.
        */
-      dinucs(complement(sequence.reverse)).reverse
+      revDinucs(complement(sequence))
     } else {
-      dinucs(sequence)
+      fwdDinucs(sequence)
     }
   }
 
-  private def dinucs(sequence: Seq[Char]): Seq[Option[(Char, Char)]] = {
-    sequence.zipWithIndex.map {
-      case (current, index) =>
-        assert(Seq('A', 'C', 'T', 'G', 'N').contains(current))
-        def previous = sequence(index - 1)
-        if (index > 0 && previous != 'N' && current != 'N') {
-          Some((previous, current))
+  private[recalibration] def fwdDinucs(sequence: String): Array[(Char, Char)] = {
+    val array = new Array[(Char, Char)](sequence.length)
+    dinucs(sequence, false, array)
+  }
+
+  private[recalibration] def revDinucs(sequence: String): Array[(Char, Char)] = {
+    val array = new Array[(Char, Char)](sequence.length)
+    dinucs(sequence, true, array)
+  }
+
+  @tailrec private def dinucs(sequence: String,
+                              swap: Boolean,
+                              array: Array[(Char, Char)],
+                              idx: Int = 0): Array[(Char, Char)] = {
+    if (idx < 0 || idx >= sequence.length) {
+      array
+    } else {
+      val current = sequence(idx)
+      // previously, this was implemented as a lookup into a set
+      // unrolling this and not using a set is 40% faster
+      //
+      // this is ugly, but as they say,
+      // "At 50, everyone has the face they deserve"
+      require(current == 'A' ||
+        current == 'C' ||
+        current == 'G' ||
+        current == 'T' ||
+        current == 'N',
+        "Saw invalid base %s. Accepted bases are A,C,G,T,N.".format(current))
```

Comment:
We've gone back and forward about this in BQSR over time (e.g., #775). IIRC, Illumina dropped the base resolution codes several years back (unambiguous bases are now just `N`), and we only recalibrate ACTG anyways. From a performance perspective, it's much cheaper to check 5 bases than to check 17.